#VRML_SIM R2023a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/objects/floors/protos/RectangleArena.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023a/projects/appearances/protos/Plastic.proto"

WorldInfo {
}
Viewpoint {
  orientation -0.5773502691896258 0.5773502691896258 0.5773502691896258 2.0944
  position 6.415651450025315 8.545717502284822 29.442143371419206
}
TexturedBackground {
}
TexturedBackgroundLight {
}
DEF area_boundary RectangleArena {
  translation 7.5 7.5 0
  floorSize 15 15
  floorTileSize 22.6 21.7
  floorAppearance Plastic {
    colorOverride 0.5 0.5 0.5
    type "padded"
    IBLStrength 0
  }
  wallThickness 0.1
  wallHeight 1
}
DEF WATER Fluid {
  translation 7.5 7.5 0.2
  children [
    DEF water Shape {
      appearance PBRAppearance {
        baseColor 0.5 0.7 0.8
        transparency 0.9
        metalness 0
      }
      geometry Box {
        size 15 15 1
      }
      castShadows FALSE
      isPickable FALSE
    }
  ]
  name "water"
  boundingObject USE water
}
DEF ship Robot {
  translation 7.8 5.7 0.6724446199258914
  children [
    DEF body Transform {
      children [
        DEF SIDE Shape {
          appearance PBRAppearance {
            baseColor 0.8 0.8 0.8
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.6 0.1 0.04
          }
        }
        DEF MAIN Shape {
          appearance PBRAppearance {
            baseColor 0.5 0.5 0.5
            roughness 1
            metalness 0
          }
          geometry Cylinder {
            height 0.1
            radius 0.25
          }
        }
      ]
    }
    DEF PROPELLER_RIGHT Propeller {
      shaftAxis 0 1 0
      centerOfThrust 0.3 -0.1 0
      thrustConstants 0.6 0
      torqueConstants 0 0
      fastHelixThreshold 72
      device RotationalMotor {
        name "propeller_motor_right"
        acceleration 10
        maxTorque 15
      }
    }
    DEF PROPELLER_LEFT Propeller {
      shaftAxis 0 1 0
      centerOfThrust -0.3 -0.1 0
      thrustConstants 0.6 0
      torqueConstants 0 0
      fastHelixThreshold 72
      device RotationalMotor {
        name "propeller_motor_left"
        acceleration 10
        maxTorque 15
      }
    }
    GPS {
    }
    Gyro {
    }
    DistanceSensor {
      translation 2.16471e-07 0 -0.02
      rotation 0 -1 0 -1.5707953071795862
      name "area_detector"
      lookupTable [
        0 0 0
        0.7 1000 0
      ]
    }
    Compass {
    }
    DistanceSensor {
      translation 0.143475 0.20043 0.02
      rotation 0 0 1 1.4399003061004252
      name "left_sensor"
      lookupTable [
        0 0 0
        1 1000 0
      ]
    }
    DistanceSensor {
      translation -0.143475 0.20043 0.02
      rotation 0 0 1 1.7016996938995748
      name "right_sensor"
      lookupTable [
        0 0 0
        1 1000 0
      ]
    }
    Receiver {
      type "serial"
      channel 5
    }
    Emitter {
      type "serial"
      channel 5
    }
    Camera {
      translation -5.0981e-08 0 -0.05
      rotation -0.7071062811861719 -3.3905013482585576e-09 0.7071072811865696 3.14159
      name "cam_underwater"
    }
  ]
  name "ship_swarm"
  contactMaterial "db"
  immersionProperties [
    ImmersionProperties {
      fluidName "water"
    }
  ]
  boundingObject Group {
    children [
      USE SIDE
      USE MAIN
    ]
  }
  physics Physics {
    density 800
    damping Damping {
      linear 0.9
    }
  }
  controller "ship_swarm_quantumGate_instant_mouvement"
  supervisor TRUE
}
DEF ship Robot {
  translation 10 6.1 0.6724446199258914
  children [
    DEF body Transform {
      children [
        DEF SIDE Shape {
          appearance PBRAppearance {
            baseColor 0.8 0.8 0.8
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.6 0.1 0.04
          }
        }
        DEF MAIN Shape {
          appearance PBRAppearance {
            baseColor 0.5 0.5 0.5
            roughness 1
            metalness 0
          }
          geometry Cylinder {
            height 0.1
            radius 0.25
          }
        }
      ]
    }
    DEF PROPELLER_RIGHT Propeller {
      shaftAxis 0 1 0
      centerOfThrust 0.3 -0.1 0
      thrustConstants 0.6 0
      torqueConstants 0 0
      fastHelixThreshold 72
      device RotationalMotor {
        name "propeller_motor_right"
        acceleration 10
        maxTorque 15
      }
    }
    DEF PROPELLER_LEFT Propeller {
      shaftAxis 0 1 0
      centerOfThrust -0.3 -0.1 0
      thrustConstants 0.6 0
      torqueConstants 0 0
      fastHelixThreshold 72
      device RotationalMotor {
        name "propeller_motor_left"
        acceleration 10
        maxTorque 15
      }
    }
    GPS {
    }
    Gyro {
    }
    DistanceSensor {
      translation 2.16471e-07 0 -0.02
      rotation 0 -1 0 -1.5707953071795862
      name "area_detector"
      lookupTable [
        0 0 0
        0.7 1000 0
      ]
    }
    Compass {
    }
    DistanceSensor {
      translation 0.143475 0.20043 0.02
      rotation 0 0 1 1.4399003061004252
      name "left_sensor"
      lookupTable [
        0 0 0
        1 1000 0
      ]
    }
    DistanceSensor {
      translation -0.143475 0.20043 0.02
      rotation 0 0 1 1.7016996938995748
      name "right_sensor"
      lookupTable [
        0 0 0
        1 1000 0
      ]
    }
    Receiver {
      type "serial"
      channel 5
    }
    Emitter {
      type "serial"
      channel 5
    }
    Camera {
      translation -5.0981e-08 0 -0.05
      rotation -0.7071062811861719 -3.3905013482585576e-09 0.7071072811865696 3.14159
      name "cam_underwater"
    }
  ]
  name "ship_swarm(1)"
  contactMaterial "db"
  immersionProperties [
    ImmersionProperties {
      fluidName "water"
    }
  ]
  boundingObject Group {
    children [
      USE SIDE
      USE MAIN
    ]
  }
  physics Physics {
    density 800
    damping Damping {
      linear 0.9
    }
  }
  controller "ship_swarm_quantumGate_instant_mouvement"
  supervisor TRUE
}
DEF Target Solid {
  translation 7.5 10 0.1
  children [
    Group {
      children [
        DEF main_area Shape {
          appearance PBRAppearance {
            baseColor 0.6 0.2 0.2
            metalness 0
          }
          geometry Cylinder {
            height 0.8
            radius 0.6
          }
          castShadows FALSE
        }
      ]
    }
  ]
  boundingObject Group {
    children [
      USE main_area
    ]
  }
}
